### PR TITLE
Allow for Transformer layers operating on densely connected graph

### PR DESCRIPTION
**Pull request by mmjb**

### CODE REVIEW

File: tf2_gnn/layers/graph_densely_connected_transformer.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    def call(self, inputs, training: bool):
+        """
+        Compute Transformer layer on graph nodes.
+
+        Used shape abbreviations:
+        V ~ num _v_ertices
+        D ~ vertex representation _d_imension
+        H ~ attention _h_ead size
+        N ~ _n_um attention heads
+        P ~ _n_number of node pairs in total, where P = sum(graph_size**2 for graph_size in batch)
```
Comment (pashminacameron) [inline tf2_gnn/layers/graph_densely_connected_transformer.py:108]: P ~ number of node _p_airs in total

File: tf2_gnn/layers/graph_densely_connected_transformer.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            value=tf.reshape(
+                node_qkv, shape=(-1, self._num_heads, 3 * self._head_size)
+            ),
+            num_or_size_splits=3,
+            axis=-1,
+        )  # 3 times [V, N, H]
+
+        # Compute attention scores between queries and keys by doing an inner product,
+        # normalised by the square root of the representation size.
+        scores = tf.einsum(
```
Comment (pashminacameron) [inline tf2_gnn/layers/graph_densely_connected_transformer.py:177]: I find it useful to keep information about which kind of attention is being implemented when there is an einsum involved in a comment, for example: `Attention(Q, K) = softmax(Q K^T/sqrt(D))` ?

File: tf2_gnn/layers/graph_densely_connected_transformer.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        self._attention_probs_dropout_rate = attention_probs_dropout_rate
+
+        # Sublayer definitions:
+        self._node_label_to_qkv_layer = tf.keras.layers.Dense(
+            units=3 * (num_heads * head_size), activation=None, use_bias=False,
+        )
+        self._attention_to_pre_intermediate_layer = tf.keras.layers.Dense(
+            units=hidden_size, activation=None, use_bias=False,
+        )
+        self._pre_intermediate_layer_norm = tf.keras.layers.LayerNormalization(epsilon=1e-12)
```
Comment (pashminacameron) [inline tf2_gnn/layers/graph_densely_connected_transformer.py:61]: Layer Norm eps is quite small, so it caught my eye. I'm sure you have a good reason for making this so much smaller than the _Attention is all you need_ case, but then it may be good to add a note here.

File: tf2_gnn/layers/graph_densely_connected_transformer.py (hunk 2 above)
Comment (mmjb) [reply tf2_gnn/layers/graph_densely_connected_transformer.py:177]: Eh, I included the comment above the einsum to clarify the attention kind, but I think what you mean is you would prefer to have it as a math-y formula as well?

File: tf2_gnn/layers/graph_densely_connected_transformer.py (hunk 3 above)
Comment (mmjb) [reply tf2_gnn/layers/graph_densely_connected_transformer.py:61]: The `1e-12` is the TF1 default; the Keras/TF2 default is `1e-3`, which is high enough that it starts making a difference. I'm not particularly married to any constant here; I just picked `1e-12` because I knew it to not be an issue. (It's only used to avoid a division by 0 at one point, so it's influence is fairly limited beyond small numerical issues)

File: tf2_gnn/layers/graph_densely_connected_transformer.py (hunk 2 above)
Comment (pashminacameron) [reply tf2_gnn/layers/graph_densely_connected_transformer.py:177]: Yes, I meant the formula, but it's not a biggy. Sorry I forgot about this one - It's easy to forget these PRs exist in amongst all the VSTS noise..

File: tf2_gnn/layers/graph_densely_connected_transformer.py (hunk 3 above)
Comment (pashminacameron) [reply tf2_gnn/layers/graph_densely_connected_transformer.py:61]: Sounds good, thanks.
